### PR TITLE
Photon vision

### PR DESCRIPTION
**Pull request by AmitBrunicki**

### CODE REVIEW

File: src/main/java/frc/robot/commands/CenterOnCube.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import com.spikes2212.command.drivetrains.commands.DriveArcadeWithPID;
+import frc.robot.subsystems.Drivetrain;
+import frc.robot.subsystems.Vision;
+
+public class CenterOnCube extends DriveArcadeWithPID {
+
+    public CenterOnCube(Drivetrain drivetrain, Vision vision) {
+        super(drivetrain, vision::photonvisionYaw, 0, 0, drivetrain.getCameraPIDSettings());
+    }
+
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/CenterOnCube.java]: remove blank line

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import edu.wpi.first.wpilibj.TimedRobot;
 import edu.wpi.first.wpilibj2.command.CommandScheduler;
 import frc.robot.subsystems.Drivetrain;
+import org.photonvision.PhotonCamera;
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/Robot.java]: remove

File: src/main/java/frc/robot/commands/CenterOnCube.java (hunk 1 above)
Comment (AmitBrunicki) [inline src/main/java/frc/robot/commands/CenterOnCube.java]: remove this blank line

File: src/main/java/frc/robot/commands/CenterOnCube.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/commands/CenterOnCube.java]: remove this blank line

File: src/main/java/frc/robot/commands/CenterOnTable.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package frc.robot.commands;
+
+import com.spikes2212.command.drivetrains.commands.DriveArcadeWithPID;
+import frc.robot.subsystems.Drivetrain;
+import frc.robot.subsystems.Vision;
+
+public class CenterOnTable extends DriveArcadeWithPID {
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/commands/CenterOnTable.java:7]: blank line at the start of the class

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private final PhotonCamera photonCamera;
+    private final Limelight limelight;
+    private static Vision instance;
+
+    public static Vision getInstance() {
+        if (instance == null) {
+            instance = new Vision();
+        }
+        return instance;
+    }
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java:20]: add a blank line after this

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import com.spikes2212.command.DashboardedSubsystem;
+import com.spikes2212.util.Limelight;
+import org.photonvision.PhotonCamera;
+import org.photonvision.targeting.PhotonPipelineResult;
+
+public class Vision extends DashboardedSubsystem {
+
+    private final PhotonCamera photonCamera;
+    private final Limelight limelight;
+    private static Vision instance;
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java]: should be final and separated from the other variables

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    public static Vision getInstance() {
+        if (instance == null) {
+            instance = new Vision();
+        }
+        return instance;
+    }
+    public Vision() {
+        super("photonvision");
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java]: call this `vision`

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        photonCamera = new PhotonCamera("photonvision");
+        limelight = new Limelight();
+    }
+
+    public double photonvisionYaw() {
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java]: getPhotonVisionYaw

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public void setDriverMode(boolean mode) {
+        photonCamera.setDriverMode(mode);
+    }
+
+    public double limelightYaw(){
+        return limelight.getHorizontalOffsetFromTargetInDegrees();
+    }
+
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java:43]: remove blank line

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public void configureDashboard() {
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java:45]: add yaw from limelight and from photon vision to this

File: vendordeps/photonlib.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+            "artifactId": "PhotonLib-java",
+            "version": "v2022.2.0"
+        },
+        {
+            "groupId": "org.photonvision",
+            "artifactId": "PhotonTargeting-java",
+            "version": "v2022.2.0"
+        }
+    ]
+}
```
Comment (AmitBrunicki) [inline vendordeps/photonlib.json]: add blank line at the end of the file

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+            return result.getBestTarget().getYaw();
+        }
+        return 0;// no target
+    }
+
+    public void setDriverMode(boolean mode) {
+        photonCamera.setDriverMode(mode);
+    }
+
+    public double limelightYaw(){
```
Comment (AmitBrunicki) [inline src/main/java/frc/robot/subsystems/Vision.java]: call this `getLimelightHorizontalOffsetFromTargetInDegrees`

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 13]
```diff
@@ -106,6 +106,8 @@ public double getRightEncoderPosition() {
         return rightEncoder.getPosition();
     }
 
+    public PIDSettings getCameraPIDSettings() { return pidSettingsCamera; }
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Conventions A-II\a7, A-II\a8

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    private final Limelight limelight;
+
+    public static Vision getInstance() {
+        if (instance == null) {
+            instance = new Vision();
+        }
+        return instance;
+    }
+
+    public Vision() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Vision.java]: get limelight and photonCamera in the constructor and give them in `getInstance()`

Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Vision.java]: private

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        super("vision");
+        photonCamera = new PhotonCamera("photonvision");
+        limelight = new Limelight();
+    }
+
+    public double getPhotonVisionYaw() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Vision.java:28]: document

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        PhotonPipelineResult result = photonCamera.getLatestResult();
+        if (result.hasTargets()) {
+            return result.getBestTarget().getYaw();
+        }
+        return 0;// no target
+    }
+
+    public void setDriverMode(boolean mode) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Vision.java:36]: ditto

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        photonCamera.setDriverMode(mode);
+    }
+
+    public double getLimelightYaw() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Vision.java:40]: ditto

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

**Comment by Yoel-Perman-Brilliant** [comment]
Your code doesn't compile, check why

**Review by OfriRosenbaum — APPROVED** [review]

**Pull request merged**